sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala

```diff
@@ -65,13 +65,38 @@ class JDBCOptions(
   // Required parameters
   // ------------------------------------------------------------
   require(parameters.isDefinedAt(JDBC_URL), s"Option '$JDBC_URL' is required.")
-  require(parameters.isDefinedAt(JDBC_TABLE_NAME), s"Option '$JDBC_TABLE_NAME' is required.")
+
   // a JDBC URL
   val url = parameters(JDBC_URL)
-  // name of table
-  val table = parameters(JDBC_TABLE_NAME)
+  val tableName = parameters.get(JDBC_TABLE_NAME)
+  val query = parameters.get(JDBC_QUERY_STRING)
```

Comment:
I think, since the `tableName` and `query` variables don't need to be exposed to other classes, can we remove them?

Btw, I feel sharing the `tableName` variable int both write/read paths makes code some complicated, so how about splitting the variable into two part: `tableOrQuery` for reading and `outputName` for writing?

e.g., https://github.com/apache/spark/commit/d62372ab0e855c359122609f1805ce83661d510e

Reply:
@maropu Thank you for taking the time to think about this throughly. A couple of questions/comments.
1) Looks like for read path we give precedence to dbtable over query. I feel its good to explicitly disallow this with a clear message in case of an ambiguity.
2) Usage of lazy here (especially to trigger errors) makes me a little nervous. Like if we want to introduce a debug statement to print the variables in side the QueryOptions class, things will not work any more, right ? Thats the reason, i had opted to check for the "invalid query option in write path" in the write function itself (i.e when i am sure of the calling context). Perhaps that how its used every where in which case it may be okay to follow the same approach here. 

I am okay with this. Lets get some opinion from @gatorsmile. Once i have the final set of comments, i will make the changes. Thanks again.

Reply:
yea, it's okay to depend on xiao's opnion.

Reply:
Another option is to follow what we are doing in another PR: https://github.com/apache/spark/pull/21247 ? We are facing the same issue there. The options are shared by both read and write paths. However, the limitations are different.